### PR TITLE
Client-side specimen ID matching

### PR DESCRIPTION
**Pull request by fedarko**
Progress on #520 / #553. (Both of those issues have a couple moving parts, but this should address what I think is the "main" thing with both.)

## Summary of changes
- Added `getSubstringMatches()`, a simple function that (given a query string and an array of strings) returns all strings in the array that include the query string
  - Added corresponding unit tests
- Added `get_active_samples()` (named analogously to the existing `get_active_studies()`). This function just gets all sample IDs associated with all active studies. (It technically returns a `Promise` that resolves to an array of sample IDs, since the requests are done asynchronously.)
  - Made some modifications in the back-end code to enable this -- in particular, exposing a `limit` parameter that allows the front-end to issue a request for *all* samples associated with a study instead of just 20.
  - Added various tests to verify that `limit`s are being handled properly. 
  - Thanks to @charles-cowart for help with this!
  - Also created `merge_sample_responses()`, which is just a few lines of code I abstracted out from `autocomplete_search_samples()` so that I could use it in both that function and in `get_active_samples()`.
- Added an "error" function in `autocomplete_search_samples()` (which didn't previously have one). This function just creates a `bootstrapAlert()` explaining the situation.
  - Also added two other error functions to the two other instances in `plateViewer.js` that use `.then()` (i.e. Promises).
  - See the commit message of https://github.com/biocore/LabControl/pull/585/commits/ed55343e6c9855d92db0c9f0383fbefdbbcdfd86 for more details about this than are probably necessary.

- `PlateViewer.modifyWell()` now uses `get_active_samples()` and `getSubstringMatches()` to perform client-side matching: given an input specimen ID, this tries to match it up with an active sample ID.
  - There are three possible results of this:
    - exactly 1 sample matches --> great, change the well contents accordingly
    - 0 samples match --> ok, this well will just be labelled "unknown"
    - 2 or more samples match --> assign the new class "indeterminate" (it's the yellow one) to this well, indicating that manual resolution is needed. **On reloading a plate, these wells will show up as red (i.e. "unknown")** -- see caveats below. 
  - The main AJAX call that `PlateViewer.modifyWell()` used to do is now done in `PlateViewer.patchWell()`. I initially did this so I could call `patchWell()` from multiple points, but this ended up not being needed. (I still like having `modifyWell()` and `patchWell()` be separate functions, though. Feel free to ask for them to be renamed :)
  - Changed the "intermediate" well legend to reflect the behavior on reloading.
- Various small fixes around the codebase (adding missing docs in process.py, removed an unused variable declaration in plate.js, added Travis/Coveralls badges to the README, etc.)

## Caveats
There are some things I still think could be improved in the proposed changes -- noting them here because most of these are sorta subjective, so if you have any feedback I'd appreciate it.
- Client-side matching is fairly quick (at least on my computer), but it could be faster if we did stuff like caching the list of active sample IDs. The golden standard here would be applying all of these changes in bulk (instead of doing everything through `modifyWell()`, which is why you see samples' statuses change in individual "blips")... but this would probably necessitate some significant restructuring of how the plating interface works.

- ~~Some of the uses of promises in the code could be handled better (e.g. throwing a bootstrapAlert if a promise fails to resolve).~~ **Update: resolved as of https://github.com/biocore/LabControl/pull/585/commits/ed55343e6c9855d92db0c9f0383fbefdbbcdfd86.**

- This is the main caveat: after spending some time going through the code that handles well classes, it seems like adding an indeterminate class to wells on the front-end is pretty simple (as is done here). However, making that class persist on the back-end (so that reloading a plate reloads the class for each well) is going to be a pretty heavy amount of work on top of what's been done here: it seems like we'd need to modify the database schema to store this information, or at least figure out a way to use a SQL query to detect *in python* which wells are indeterminate. Took a few stabs at this last night; didn't get very far.
  - To complicate things, indeterminate wells are technically "unknown", since they don't have a valid value. From what I can tell, making these wells not "unknown" seems like it'd involve a lot of going through the code to modify things.
  - For now, the behavior is that (as mentioned above) the indeterminate class isn't preserved when reloading a plate—indeterminate cells will just show up on a reload as "unknown", which is *technically* kinda accurate since they don't contain exact sample IDs. Although technically accurate, this is fairly undesirable. (As mentioned, I updated the indeterminate legend to match, so that users understand why these cells change on reloading.)
  - I'm sure making the indeterminate class persistent is doable—and I'm happy to work when I get back next week on this—but I'm not sure that the effort is worth it, especially since the yellow cells are an indicator of "hey you need to resolve this manually". Of course if you disagree that is fine, I'm not the expert here :)

Anyway—sorry if this PR isn't comprehensive enough; I mainly wanted to make sure I had this ready to show you all where this is at today. Let me know what you think!

Screen Shot 2019-08-30 at 8 34 02 AM

### CODE REVIEW

**Comment by fedarko** [comment]
As of now, I think this is ok to be reviewed—now that error handling is added in, the two remaining "caveats" are general, subjective issues that I'd like feedback on.

**Comment by AmandaBirmingham** [comment]
@fedarko Hiya--I'm stepping into Charlie's shoes for the moment :)  I see that the Travis build for this PR did not succeed due to three failing unit tests.  Please modify the PR to address these failures; I can do a review on it once the build passes.

Thanks for all the effort you are putting in on LabControl!

**Comment by fedarko** [comment]
Thanks so much @AmandaBirmingham! This was due to me not updating tests re: the backend changes made in this PR. I updated the tests (and added some tests that the `limit` parameter sent to `Study.samples()` is being handled/validated properly), and things seem to be working on Travis now.

**Comment by fedarko** [comment]
Something potentially gnarly, but likely out of the scope of this PR... I ran ea536f3's build twice ([this one](https://travis-ci.org/biocore/LabControl/builds/581932706)), and it passed the first time but failed on the two "download" tests (`test_download` and `test_get_download_normalization_process_handler`) again.

Since these failures were both caused by `AssertionError: Async operation timed out after 10 seconds`, it seems like this is an intermittent error due to network problems (???). Not sure how best to handle this, but as a brute-force solution using `travis_retry` with the python tests might work.

**Comment by gwarmstrong** [comment]
> Something potentially gnarly, but likely out of the scope of this PR... I ran [ea536f3](https://github.com/biocore/LabControl/commit/ea536f3faf559bce9b7c440b5ac6b493fce40de6)'s build twice ([this one](https://travis-ci.org/biocore/LabControl/builds/581932706)), and it passed the first time but failed on the two "download" tests (`test_download` and `test_get_download_normalization_process_handler`) when run again.
> 
> Since these failures were both caused by `AssertionError: Async operation timed out after 10 seconds`, it seems like this is an intermittent error due to network problems (???). Not sure how best to handle this, but as a brute-force solution using `travis_retry` with the python tests might work.

You're right, this has been a recurrent problem I've noticed in the past couple weeks. I think there are a couple things we can do to try to reduce this.
* Increase the `timeout` limit in the line below:
https://github.com/biocore/LabControl/blob/1427ad163682ff11b1086e0e3b1010ba1a2166ee/labcontrol/gui/testing.py#L71
* add retry functionality _within_ the test, e.g. something like [this](https://stackoverflow.com/questions/31529151/python-3-urllib-urlretrieve-retry-on-download-fail):
```
class TestDownloadLibraryPrepShotgunProcessHandler(TestHandlerBase):
    def test_download(self):
        retries = 0
        response = None
        while retries < 5:
            try:
                response = self.get(
                    '/process/library_prep_shotgun/%d/echo_pick_list' % 1)
            except AssertionError:
                print("Error downloading on try {}.".format(retries))
            else:
                break
        if response is None:
           raise AssertionError("Async operation timed out after maximum number of retries.")

        self.assertNotEqual(response.body, '')
        self.assertTrue(response.body.startswith(
            b'Sample ID\tSource Plate Name\t'))
        self.assertEqual(response.headers['Content-Disposition'],
                         "attachment; filename=2017-10-25_"
                         "Test_compressed_gDNA_plates_1-4_indices.txt")

```

**Review by AmandaBirmingham — CHANGES_REQUESTED** [review]
Lots of good stuff here--thank you!  A few change requests, all pretty minor, I think.

File: .coveragerc
Comment (fedarko) [reply .coveragerc]: Yeah, we need both. Essentially this file is saying "don't include coverage information for these directories in the total summary"—it was already excluding the `tests/` directory (i.e. `labcontrol/db/tests/`, but it wasn't excluding the `test/` directory for the front-end (`labcontrol/gui/test/`). This change should fix that problem, and the result is that coveralls should now provide a more realistic representation of the code coverage.

...It might also be worth renaming either of the test directories to make them consistent, since having the test directories be named slightly differently is a little bit confusing :)

Comment (gwarmstrong) [reply .coveragerc]: > ...It might also be worth renaming either of the test directories to make them consistent, since having the test directories be named slightly differently is a little bit confusing :)

I agree, opening an issue.

File: labcontrol/gui/js_tests/labcontrol_tests.js
Comment (fedarko) [reply labcontrol/gui/js_tests/labcontrol_tests.js]: Honestly, I don't actually remember why I went with propEqual here. It [looks like](https://api.qunitjs.com/assert/propEqual) the differences between it and deepEqual mostly have to do with actual JS objects (i.e. not basic arrays like the ones here)—I can totally switch these asserts over to deepEqual if you'd like.

File: labcontrol/db/process.py
Comment (fedarko) [reply labcontrol/db/process.py]: Understood and agreed! Just pushed a fix that changes this.

File: labcontrol/gui/static/js/labcontrol.js
Comment (fedarko) [reply labcontrol/gui/static/js/labcontrol.js]: I'm ok with that (so long as we're ok with only supporting ES6-capable browsers). To clarify, do you just mean using `let` instead of `var` from here on, or actively porting the old code to use `let`?

`grep -ri "var " labcontrol/gui/static/js/ labcontrol/gui/js_tests/*.js | wc -l` finds 171 hits and `grep -ri "let " labcontrol/gui/static/js/ labcontrol/gui/js_tests/*.js | wc -l` finds 28 hits, so the latter option might take some time :)

File: labcontrol/gui/static/js/plateViewer.js
Comment (fedarko) [reply labcontrol/gui/static/js/plateViewer.js]: Yeah—I wrote this comment before I realized that my understanding of how LabControl handles multiple studies was inaccurate. Just pushed a commit that should address this, thanks!

File: labcontrol/gui/static/js/plateViewer.js
Comment (fedarko) [reply labcontrol/gui/static/js/plateViewer.js]: `possiblyNewContent` might have been a bad choice for the variable name, sorry—even if this code fails to find a one-to-one match, it still updates the database with the `content` that was passed in (which also should be "new," since `modifyWell()` is called when a cell changes). I'm open to renaming this / adding comments to improve readability.

File: labcontrol/gui/templates/plate.html
Comment (fedarko) [reply labcontrol/gui/templates/plate.html]: I hadn't noticed the duplicate IDs, but I agree that that is egregious. Just pushed a commit that should fix that particular wart, at least.

It looks like duplicating the text is needed (at least with the way the HTML/CSS is written currently) because truncating the hidden text messes with the layout of the next section ("Well Comments")—

Screen Shot 2019-09-09 at 7 21 20 PM

For comparison, here's a screenshot with the duplicated text solution in use—

Screen Shot 2019-09-09 at 7 24 53 PM

I changed the hidden divs to be visible for these screenshots, but the effects shown here persist regardless of visibility.

Anyway! Yeah, entirely agreed that the need for this hidden section at all is a hack (and I hereby take responsibility for that, since I approved this code when it was merged in :). Removing the duplicate IDs makes it more tolerable, but it'd definitely be a good idea to adjust this to make the bottom row expand without needing a hidden section. I just created #590 to address this (can try to fix that in this PR if you'd like, but I don't think it's super urgent).

**Comment by fedarko** [comment]
@AmandaBirmingham Thank you so much for the many helpful comments! I'm not going to be able to respond to them all today, but I'll get to work on hopefully addressing all of these tomorrow.

File: .coveragerc
Comment (fedarko) [reply .coveragerc]: Marking this as resolved—we can follow up with this in #589.

File: labcontrol/db/study.py
Comment (fedarko) [reply labcontrol/db/study.py]: Should be resolved in latest commit.

File: labcontrol/db/tests/test_study.py
Comment (fedarko) [reply labcontrol/db/tests/test_study.py]: Yep, now that `Study.samples()` accepts strings as input the `"1"` case would pass.

Something a little bit funky with the casting approach is that `int(1.0)` works (gets truncated to 1), but `int("1.0")` fails. Since (as far as I can tell) the only use of `Study.samples()` in LabControl involves passing in strings as inputs, this just means that if a request is sent with e.g. `...&limit=1.0` then that request will fail (as should be expected).

File: labcontrol/gui/static/js/labcontrol.js
Comment (fedarko) [reply labcontrol/gui/static/js/labcontrol.js]: Thanks!

File: labcontrol/gui/static/js/plateViewer.js
Comment (fedarko) [reply labcontrol/gui/static/js/plateViewer.js]: Done: #591 now describes the issue.

File: labcontrol/gui/static/js/plateViewer.js
Comment (fedarko) [reply labcontrol/gui/static/js/plateViewer.js]: Good catch, thank you!

File: labcontrol/gui/static/js/plateViewer.js
Comment (fedarko) [reply labcontrol/gui/static/js/plateViewer.js]: Got it. I reworded the three uses of bootstrapAlerts in this commit to specify the function the alerts are being thrown from.

File: labcontrol/gui/static/js/plateViewer.js
Comment (fedarko) [reply labcontrol/gui/static/js/plateViewer.js]: Renamed `possiblyNewContent` to `possiblyMatchedContent`, in the hopes of making this a bit clearer. Can do more if needed.

Comment (fedarko) [reply labcontrol/gui/static/js/plateViewer.js]: Renamed `possiblyNewContent` to `possiblyMatchedContent`, in the hopes of making this a bit clearer. Can do more if needed.

File: labcontrol/gui/static/js/plateViewer.js
Comment (fedarko) [reply labcontrol/gui/static/js/plateViewer.js]: Created #592 accordingly.

File: labcontrol/gui/static/js/plateViewer.js
Comment (fedarko) [reply labcontrol/gui/static/js/plateViewer.js]: Understood—apologies for the naive comment. I've updated the comments accordingly, and now #173 is mentioned in this part of the code.

File: labcontrol/gui/js_tests/labcontrol_tests.js
Comment (fedarko) [reply labcontrol/gui/js_tests/labcontrol_tests.js]: For consistency's sake, the tests now use `assert.deepEqual`.

**Comment by fedarko** [comment]
@AmandaBirmingham I think I've addressed all of your comments here (I left a few comments "unresolved" where I wasn't sure if you'd like me to do more).

Please let me know if you'd like me to do anything else before we merge this in—thanks again!

File: labcontrol/db/study.py
Comment (AmandaBirmingham) [inline labcontrol/db/study.py]: Is there a reason to limit this try/catch to particular errors?  Since the only statement wrapped in the try/catch is the cast, and since we aren't returning different error messages based on different error classes, I think it is safe to catch all errors (with a simple `except:`) and raise your custom ValueError.

**Review by AmandaBirmingham — APPROVED** [review]
Just one minor comment that I don't think requires action right now. Ready to merge after a second reviewer approves?

File: labcontrol/db/study.py
Comment (fedarko) [reply labcontrol/db/study.py]: Update—just pushed a commit that changes this to `except Exception:` (PEP8 apparently recommends this form over `except:`, since the latter catches stuff like KeyboardInterrupts).

**Comment by fedarko** [comment]
Since @gwarmstrong is out right now, I suppose I can request a review from someone else?

**Comment by fedarko** [comment]
@qiyunzhu and I just sat down and walked through the code changes, and he has given his verbal approval of this PR.

While going over the code I noticed one small case that isn't explicitly tested in the `Study.samples()` unit test (strings of negative ints), so I added a check for that in the latest commit.

Anyway, once those tests pass (or at least fail only due to the intermittent and unrelated download error mentioned in #588), I can merge this PR in.

**Comment by fedarko** [comment]
Build passed—merging in now. Thanks @AmandaBirmingham and @qiyunzhu for looking over this!

**Pull request merged**
